docs: react 18, streaming SSR, rsc with new apis (#33986)
vercel/next.js: docs/api-reference/next/streaming.md

---
description: Streaming related APIs to build Next.js apps in streaming SSR or with React Server Components.
---

# next/streaming

> The `next/streaming` module provides streaming related APIs to port the existing functionality of Next.js app to streaming scenarios or assign users ability to interact much easier with React Server Components.

Reviewer: Maybe mention that these are experimental?

```suggestion
The experimental `next/streaming` module provides streaming related APIs to port the existing functionality of Next.js apps to streaming scenarios and facilitate the usage of React Server Components.
```

## unstable_useWebVitalsReport

> Next.js provides an `_app` component level function, [`reportWebVitals`](docs/advanced-features/measuring-performance), for tracking performance metrics. With server components, you may have a pure server-side custom `_app` component which doesn't run client effects, so this existing API won't work.

Reviewer 2: ```suggestion
Next.js provides an `_app` component-level function, [`reportWebVitals`](docs/advanced-features/measuring-performance), for tracking performance metrics. With Server Components, you may have a pure server-side custom `_app` component (which doesn't run client effects) so the existing API won't work.
```

> Next.js provides an App component level function `reportWebVitals` for tracking performance metrics, checkout [measuring-performance](docs/advanced-features/measuring-performance) for more details. With moving to server components, you might have a pure server side custom `_app` component which doesn't run client effects.

Reviewer: ```suggestion
Next.js provides an `App` component level function, [`reportWebVitals`](docs/advanced-features/measuring-performance), for tracking performance metrics. With server components, you may have a pure server-side custom `_app` component which doesn't run client effects, so this existing API won't work.
```

> With the new `unstable_useWebVitalsReport` API, you're able to track web vitals metrics in client components.

Reviewer 2: ```suggestion
With the new `unstable_useWebVitalsReport` API, you're able to track [Core Web Vitals](https://nextjs.org/learn/seo/web-performance) in client components:
```

> With this API, you're able to track web vitals metrics through this hook in client components.

Reviewer: To clarify which one "this API" refers to:

```suggestion
With the new `unstable_useWebVitalsReport` API, you're able to track web vitals metrics in client components.
```

```jsx
// pages/_app.js
import { unstable_useWebVitalsReport } from 'next/streaming'

export default function Home() {
  unstable_useWebVitalsReport((data) => {
    console.log(data)
  })

  return <div>Home Page</div>
}
```

> This method could also be used to replace static exported `reportWebVitals` functions in your existing `_app`.

Reviewer 2: ```suggestion
This method could also be used to replace statically exported `reportWebVitals` functions in your existing `_app`:
```

> Or it could also be an alternative way for you to replace static exported `reportWebVitals` function in your existing \_app.

Reviewer: ```suggestion
This method could also be used to replace static exported `reportWebVitals` functions in your existing `_app`.
```

```jsx
// pages/_app.server.js
import Layout from '../components/layout.client.js'

export default function App({ children }) {
  return <Layout>{children}</Layout>
}
```

```jsx
// components/layout.client.js

> import { unstable_useWebVitalsReport as useWebVitalsReport } from 'next/streaming'
> 
> export default function Layout() {
>   useWebVitalsReport((data) => {
>     console.log(data)

Reviewer 2: ```suggestion
import { unstable_useWebVitalsReport } from 'next/streaming'

export default function Layout() {
  unstable_useWebVitalsReport((data) => {
    console.log(data)
```

  })

  return (
    <div className="container">
      <h1>Hello</h1>
      <div className="main">{children}</div>
    </div>
  )
}
```

## unstable_useRefreshRoot

> Since server components are rendered on the server side when requesting to server, in some cases you might need to partially refresh server rendered content.

Reviewer 2: ```suggestion
Since Server Components are rendered on the server-side, in some cases you might need to partially refresh content from the server.
```

> Since server components are rendered on server side when requesting to server, in some case you might need to partially refresh server rendered content.

Reviewer: ```suggestion
Since server components are rendered on the server side when requesting to server, in some cases you might need to partially refresh server rendered content.
```

> For instance, let's say you build a search bar in a client component and display search results through server components. You'd want to update the search results while typing and re-render the results list with a certain frequency (e.g. with each keystroke or on a debounce).

Reviewer 2: ```suggestion
For example, a search bar (client component) which displays search results as server components. You'd want to update the search results while typing and rerender the results list with a certain frequency (e.g. with each keystroke or on a debounce).
```

> For instance, you build a search bar in client component and display few search results by server components, you'd like to update the search results while typing. Then you'd like to re-render the results list in a certain frequency, it could be each typing, or be batched by debounce.

Reviewer: ```suggestion
For instance, let's say you build a search bar in a client component and display search results through server components. You'd want to update the search results while typing and re-render the results list with a certain frequency (e.g. with each keystroke or on a debounce).
```

> The `unstable_useRefreshRoot` hook returns a `refresh` API to let you re-render the React tree smoothly without flickering. This is only allowed to be used on the client side and will only affect server components at the moment.

Reviewer 2: ```suggestion
The `unstable_useRefreshRoot` hook returns a `refresh` API to let you re-render the React tree smoothly without flickering. This is only allowed for use on the client-side and will only affect Server Components at the moment.
```

> The hook `unstable_useRefreshRoot` returns a `refresh` API to let you rerender the React tree smoothly without flicking. This is only allowed to be used on client side and will only effect server components so far.

Reviewer: ```suggestion
The `unstable_useRefreshRoot` hook returns a `refresh` API to let you re-render the React tree smoothly without flickering. This is only allowed to be used on the client side and will only affect server components at the moment.
```
